Add support for error-free reads with --errfree argument and separate BAM file processing

teojcryan opened the pull request:
This pull request adds support for generating error-free reads using the --errfree argument. Key changes include:

- Added --errfree argument: Allows users to generate SAM files without sequencing errors.
- Separate error-free BAM file processing: Introduced distinct rules to handle error-free BAM files throughout the pipeline, ensuring separation from standard reads.
- New output directory for error-free BAM files: Error-free BAM files are stored separately to maintain a clear distinction.

farchaab commented on mess/workflow/rules/processing/reads.smk:
```diff
@@ -537,6 +540,146 @@ if not SKIP_SHUFFLE:
             """
 
+if ERRFREE:
+    rule fix_art_sam_ef:
```
We use the fix_art_sam rule to fix #22. Since error free SAM files do not have sequencing errors, we don't need this rule

farchaab commented on mess/workflow/rules/processing/reads.smk:
```diff
@@ -537,6 +540,146 @@ if not SKIP_SHUFFLE:
+            {input} > {output}
+            """
+    
+    rule convert_sam_to_bam_ef:
```
I simplified the rules below in the [readproc branch](https://github.com/metagenlab/MeSS/tree/readproc), It would be good to adapt the PR according to that branch !

farchaab commented on mess/workflow/rules/processing/reads.smk:
I would remove the fixing of error free SAM files

farchaab commented:
I made some changes on how `mess` processes reads and alignments in [readproc](https://github.com/metagenlab/MeSS/tree/readproc). To summarize, I simplified alignments and reads concatenation (It was previously concatenating contigs per fasta and then all fastas per sample, now it does it in one go).

Could you try to merge into readproc and try to solve conflicts ?

farchaab commented:
> Thanks @farchaab! I'm giving it a go to push my changes into readproc. I'm running into an issue with `cat_fastqs` rule, specifically with the `aggregate()` function.
> 
> ```
> InputFunctionException in rule cat_fastqs in file path/MeSS/mess/workflow/rules/processing/reads.smk, line 343:
> Error:
>   AttributeError: 'Series' object has no attribute 'itertuples'
> Wildcards:
>   sample=sample1
>   p=1
> Traceback:
>   File "path/MeSS/mess/workflow/rules/processing/reads.smk", line 345, in <lambda>
>   File "path/MeSS/mess/workflow/rules/preflight/functions.smk", line 152, in aggregate
>   File "/rds/projects/w/wheelern-wwmgs/.conda/envs/mess/lib/python3.12/site-packages/pandas/core/generic.py", line 6299, in __getattr__ (rule cat_fastqs, line 648, path/MeSS/mess/workflow/rules/processing/reads.smk)
> ```

Hello @teojcryan, this should be fixed with 926e0025a3e0582d8c98eb77ddc1d970beb47ce5

teojcryan commented:
Thanks @farchaab! I think it should be working now - I've updated the ef variant of the rules to match the regular rules, and also removed `fix_art_sam_ef`.
